ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -191,80 +273,272 @@ static INT_PTR CALLBACK ShowMessageExtProc(HWND hwndDlg, UINT uMsg, WPARAM wPara
+  fname = remove_slash(fname);
+
+  tstring tstr_filter = widen(filter);
+  replace(tstr_filter.begin(), tstr_filter.end(), '|', '\0');
+  tstring tstr_fname = widen(fname);
+  tstring tstr_dir = widen(dir);
+  tstring tstr_title = widen(title);
+
+  wchar_t wstr_fname1[4096];
+  wcsncpy_s(wstr_fname1, tstr_fname.c_str(), 4096);
 
-void message_position(int x, int y) {
+  ZeroMemory(&ofn, sizeof(ofn));
+  ofn.lStructSize = sizeof(ofn);
+  ofn.hwndOwner = enigma::hWnd;
+  ofn.lpstrFile = wstr_fname1;
+  ofn.nMaxFile = 4096;
+  ofn.lpstrFilter = tstr_filter.c_str();
+  ofn.nFilterIndex = 0;
+  ofn.lpstrTitle = tstr_title.c_str();
+  ofn.lpstrInitialDir = tstr_dir.c_str();
+  ofn.Flags = OFN_EXPLORER | OFN_FILEMUSTEXIST | OFN_HIDEREADONLY | OFN_NOCHANGEDIR | OFN_ALLOWMULTISELECT;
+
+  if (GetOpenFileNameW(&ofn) != 0) {
+    tstring tstr_fname1 = wstr_fname1;
+    tstr_fname1 += '\\';
+
+    size_t pos = 0;
+    size_t prevlen = pos;
+    size_t len = wcslen(wstr_fname1);
+
+    while (pos < len) {
+      if (wstr_fname1[len - 1] != '\n' && wstr_fname1[len] == '\0')
+        wstr_fname1[len] = '\n';
+
+      prevlen = len;
+      len = wcslen(wstr_fname1);
+      pos += (len - prevlen) + 1;
+    }
 
+    tstring tstr_fname2 = wstr_fname1;
+    if (tstr_fname2[len - 1] == '\n') tstr_fname2[len - 1] = '\0';
+    if (tstr_fname2[len - 2] == '\\') tstr_fname2[len - 2] = '\0';
+    tstr_fname2 = tstring_replace_all(tstr_fname2, L"\n", L"\n" + tstr_fname1);
+    size_t rm = tstr_fname2.find_first_of(L'\n');
+
+    if (rm != string::npos)
+      tstr_fname2 = tstr_fname2.substr(rm + 1, tstr_fname2.length() - (rm + 1));
+
+    tstr_fname2.append(L"\0");
+    if (tstr_fname2.length() >= 4095) {
+      tstr_fname2 = tstr_fname2.substr(0, 4095);
+      size_t end = tstr_fname2.find_last_of(L"\n");
+      tstr_fname2 = tstr_fname2.substr(0, end);
+      tstr_fname2.append(L"\0");
+    }
+
+    return string_replace_all(shorten(tstr_fname2), "\\\\", "\\");
+  }
+
+  return "";
 }
 
-void message_size(int w, int h) {
+static inline string get_save_filename_helper(string filter, string fname, string dir, string title) {
+  OPENFILENAMEW ofn;
+
+  filter = string(filter).append("||");
+  fname = remove_slash(fname);
 
+  tstring tstr_filter = widen(filter);
+  replace(tstr_filter.begin(), tstr_filter.end(), '|', '\0');
+  tstring tstr_fname = widen(fname);
+  tstring tstr_dir = widen(dir);
+  tstring tstr_title = widen(title);
+
+  wchar_t wstr_fname[MAX_PATH];
+  wcsncpy_s(wstr_fname, tstr_fname.c_str(), MAX_PATH);
+
+  ZeroMemory(&ofn, sizeof(ofn));
+  ofn.lStructSize = sizeof(ofn);
+  ofn.hwndOwner = enigma::hWnd;
+  ofn.lpstrFile = wstr_fname;
+  ofn.nMaxFile = MAX_PATH;
+  ofn.lpstrFilter = tstr_filter.c_str();
+  ofn.nFilterIndex = 0;
+  ofn.lpstrTitle = tstr_title.c_str();
+  ofn.lpstrInitialDir = tstr_dir.c_str();
+  ofn.Flags = OFN_EXPLORER | OFN_PATHMUSTEXIST | OFN_HIDEREADONLY | OFN_OVERWRITEPROMPT | OFN_NOCHANGEDIR;
+
+  if (GetSaveFileNameW(&ofn) != 0)
```

Comment:
This block is identical to `get_open_filename_helper`, except it calls `GetSaveFileNameW` instead of `GetSaveFileNameW`.

Reply:
Ok, I think I finally fixed it.